File: gate/static/promo-clock.js
/**
 * Promo Clock — fail-closed marketing dates for liveness surfaces.
 * Past "Next drill" must not render as upcoming.
 *
 * Usage:
 *   <div class="drill-bar" data-promo-clock
 *        data-next-at="2026-08-18T17:00:00-07:00"
 *        data-last-proved-at="2026-08-24T06:31:00Z"
 *        data-label="AWS Loft SF"
 *        data-href="/fuse/fuse_velaru_drill"></div>
 *   <script src="https://gate.velaru.xyz/static/promo-clock.js" defer></script>
 *
 * Or set window.PROMO_CLOCK = { next_at, last_proved_at, label, href }
 * before load; targets #drill-bar by default.
 */
(function () {
  "use strict";

  function parseTs(s) {
    if (!s || !String(s).trim()) return null;
    var t = String(s).trim();
    if (/^\d{4}-\d{2}-\d{2}$/.test(t)) t += "T23:59:59Z";
    var d = new Date(t);
    return isNaN(d.getTime()) ? null : d;
  }

  function fmt(d) {
    try {
      return d.toLocaleString(undefined, {
        weekday: "short",
        month: "short",
        day: "numeric",
        year: "numeric",
        hour: "numeric",
        minute: "2-digit",
        timeZoneName: "short",
      });
    } catch (e) {
      return d.toISOString();
    }
  }

  function evaluate(cfg, now) {
    now = now || new Date();
    var next = parseTs(cfg.next_at);
    var proved = parseTs(cfg.last_proved_at);
    if (cfg.next_at && !next) {
      return { mode: "invalid", render: false };
    }
    if (next && next.getTime() > now.getTime()) {
      return {
        mode: "upcoming",
        render: true,
        headline: "Next drill",
        when: next,
        label: cfg.label,
        href: cfg.href,
      };
    }
    if (proved) {
      return {
        mode: "proved",
        render: true,
        headline: "Last proved",
        when: proved,
        label: cfg.label,
        href: cfg.href,
        stale: !!(next && next.getTime() <= now.getTime()),
      };
    }
    return {
      mode: "hidden",
      render: false,
      stale: !!(next && next.getTime() <= now.getTime()),
    };
  }

  function apply(el, state) {
    if (!el) return;
    if (!state.render) {
      el.hidden = true;
      el.setAttribute("data-promo-mode", state.mode || "hidden");
      el.innerHTML = "";
      return;
    }
    el.hidden = false;
    el.setAttribute("data-promo-mode", state.mode);
    var parts = ["<strong>" + state.headline + "</strong>"];
    if (state.label) parts.push(escapeHtml(state.label));
    if (state.when) parts.push(escapeHtml(fmt(state.when)));
    var html = parts.join(" · ");
    if (state.href) {
      html += ' · <a href="' + escapeAttr(state.href) + '">open →</a>';
    }
    el.innerHTML = html;
  }

  function escapeHtml(s) {
    return String(s)
      .replace(/&/g, "&amp;")
      .replace(/</g, "&lt;")
      .replace(/>/g, "&gt;")
      .replace(/"/g, "&quot;");
  }
  function escapeAttr(s) {
    return escapeHtml(s).replace(/'/g, "&#39;");
  }

  function cfgFromEl(el) {
    return {
      next_at: el.getAttribute("data-next-at") || el.getAttribute("data-promo-next"),
      last_proved_at:
        el.getAttribute("data-last-proved-at") ||
        el.getAttribute("data-promo-proved"),
      label: el.getAttribute("data-label") || el.getAttribute("data-promo-label"),
      href: el.getAttribute("data-href") || el.getAttribute("data-promo-href"),
    };
  }

  function run() {
    var nodes = document.querySelectorAll("[data-promo-clock]");
    if (!nodes.length) {
      var bar = document.getElementById("drill-bar");
      if (bar) {
        var globalCfg = window.PROMO_CLOCK || {};
        // Prefer data-* on the bar if present; else global; else parse legacy text (last resort: hide if Aug past)
        var cfg = Object.assign({}, globalCfg, cfgFromEl(bar));
        if (!cfg.next_at && !cfg.last_proved_at) {
          // Legacy hardcoded Check banner: fail closed if it still says a past calendar date
          var text = (bar.textContent || "").toLowerCase();
          var m = text.match(
            /(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)[a-z]*\s+\d{1,2}\s+\d{4}/i
          );
          if (m) {
            try {
              var guessed = new Date(m[0] + " 23:59:59 UTC");
              if (!isNaN(guessed.getTime()) && guessed.getTime() < Date.now()) {
                // Prefer last restraint feed item if page exposed RECENT_BLOCKS
                if (
                  window.RECENT_BLOCKS &&
                  window.RECENT_BLOCKS[0] &&
                  window.RECENT_BLOCKS[0].issued_at
                ) {
                  cfg.last_proved_at = window.RECENT_BLOCKS[0].issued_at;
                  cfg.label = cfg.label || "public restraint";
                  cfg.href =
                    cfg.href ||
                    (window.RECENT_BLOCKS[0].fuse_id
                      ? "/fuse/" + window.RECENT_BLOCKS[0].fuse_id
                      : null);
                } else {
                  apply(bar, { mode: "hidden", render: false, stale: true });
                  return;
                }
              }
            } catch (e) {
              apply(bar, { mode: "hidden", render: false });
              return;
            }
          }
        }
        apply(bar, evaluate(cfg));
      }
      return;
    }
    nodes.forEach(function (el) {
      apply(el, evaluate(cfgFromEl(el)));
    });
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", run);
  } else {
    run();
  }

  window.PromoClock = { evaluate: evaluate, run: run };
})();

File: gate/promo_clock.py
"""Promo Clock — fail-closed surface dates for liveness products.

Marketing clocks must not lie. If \"next drill\" is in the past, do not
show it as upcoming — flip to last-proved or hide.

Product cousin: Stale LIVE Rejector (clearance). This is Stale Promo Rejector (site).
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

SPEC = "gate-promo-clock-v1"


def _parse(ts: str | None) -> datetime | None:
    if not ts or not str(ts).strip():
        return None
    t = str(ts).strip().replace("Z", "+00:00")
    # allow "2026-08-18T17:00:00-07:00" and date-only
    if len(t) == 10 and t[4] == "-" and t[7] == "-":
        t = t + "T23:59:59+00:00"
    try:
        dt = datetime.fromisoformat(t)
    except ValueError:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def evaluate(
    *,
    next_at: str | None = None,
    last_proved_at: str | None = None,
    label: str | None = None,
    href: str | None = None,
    now: str | None = None,
) -> dict[str, Any]:
    """Return a fail-closed promo surface state.

    modes:
      upcoming — next_at in the future (safe to say \"Next drill\")
      proved   — next is stale/missing; last_proved_at exists → \"Last proved\"
      hidden   — nothing honest to show (do not render a dead Next)
      invalid  — unparseable timestamps (treat as hidden)
    """
    now_dt = _parse(now) or datetime.now(timezone.utc)
    nxt = _parse(next_at)
    proved = _parse(last_proved_at)

    if next_at and nxt is None:
        return {
            "spec": SPEC,
            "ok": False,
            "mode": "invalid",
            "render": False,
            "headline": None,
            "reason": "next_at_unparseable",
            "now": _iso(now_dt),
        }

    if nxt is not None and nxt > now_dt:
        return {
            "spec": SPEC,
            "ok": True,
            "mode": "upcoming",
            "render": True,
            "headline": "Next drill",
            "when": _iso(nxt),
            "when_display": next_at,
            "label": label,
            "href": href,
            "stale": False,
            "now": _iso(now_dt),
        }

    # next missing or past → never show as Next
    if proved is not None:
        return {
            "spec": SPEC,
            "ok": True,
            "mode": "proved",
            "render": True,
            "headline": "Last proved",
            "when": _iso(proved),
            "when_display": last_proved_at,
            "label": label,
            "href": href,
            "stale": bool(nxt is not None and nxt <= now_dt),
            "stale_next_at": _iso(nxt) if nxt else None,
            "now": _iso(now_dt),
        }

    return {
        "spec": SPEC,
        "ok": True,
        "mode": "hidden",
        "render": False,
        "headline": None,
        "reason": "stale_or_missing_next_without_proved",
        "stale": bool(nxt is not None and nxt <= now_dt),
        "stale_next_at": _iso(nxt) if nxt else None,
        "label": label,
        "href": href,
        "now": _iso(now_dt),
    }


def manifest(public_url: str = "", **kwargs: Any) -> dict[str, Any]:
    base = (public_url or "").rstrip("/")
    state = evaluate(**kwargs) if kwargs else evaluate()
    return {
        "spec": SPEC,
        "name": "Promo Clock",
        "one_liner": "Past \"next drill\" dates fail closed — show last proved or hide.",
        "cousin": "stale_live — clearance; promo_clock — site surface",
        "well_known": f"{base}/.well-known/promo-clock.json" if base else None,
        "script": f"{base}/static/promo-clock.js" if base else None,
        "usage": {
            "html": (
                '<div class="drill-bar" data-promo-clock '
                'data-next-at="2026-09-01T17:00:00-07:00" '
                'data-last-proved-at="2026-08-24T06:31:00Z" '
                'data-label="AWS Loft SF" data-href="/fuse/demo"></div>'
                '<script src="/static/promo-clock.js" defer></script>'
            ),
            "rule": "If next_at <= now → never headline Next drill.",
        },
        "state": state,
    }
